### PR TITLE
OIDC Client

### PR DESCRIPTION
**Pull request by teemukataja**
[Build Status](https://travis-ci.org/CSCfi/oidc-client)
[Coverage Status](https://coveralls.io/github/CSCfi/oidc-client?branch=oidc-client)
[Documentation Status](https://csc-oidc-client.readthedocs.io/en/latest/?badge=latest)

To do on this PR:
- [x] Write `README.md`
- [x] Write docs

Postponed till later PRs:
- Finish unit tests (100%)
- Create integration tests

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+Callback
+~~~~~~~~
+
+The callback endpoint ``/callback`` acts as a landing site for the returning user from the AAI server.
+Upon returning to the OIDC Client from the AAI server, OIDC Client extracts ``state`` and ``code`` from the callback request,
+and uses these values to request a token from the AAI server. Upon a successful retrieval of an access token, the access token
+is saved to the browser cookies.
+
+Other cookies created cookies include unsafe cookies for the purpose of displaying values in UI for logged in state and bona fide status.
```
Comment (blankdots) [inline docs/api.rst]:
```suggestion
Some of the created cookies can be considered _unsafe_ (not `http_only`) for the purpose of displaying values in UI for logged in state and bona fide status.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## OIDC Client
+
+[![Build Status](https://travis-ci.org/CSCfi/oidc-client.svg?branch=master)](https://travis-ci.org/CSCfi/oidc-client)
+[![Coverage Status](https://coveralls.io/repos/github/CSCfi/oidc-client/badge.svg)](https://coveralls.io/github/CSCfi/oidc-client)
+[![Documentation Status](https://readthedocs.org/projects/csc-oidc-client/badge/?version=latest)](https://csc-oidc-client.readthedocs.io/en/latest/?badge=latest)
```
Comment (blankdots) [inline README.md]:
```suggestion
[![Documentation Status](https://readthedocs.org/projects/csc-oidc-client/badge/?version=latest)](https://csc-oidc-client.readthedocs.io/en/latest/?badge=latest)

CSC OIDC Client is a lightweight [aiohttp](https://aiohttp.readthedocs.io/en/stable/) web application used for interacting with OIDC servers.
The source code is delivered with [ELIXIR AAI](https://elixir-europe.org/services/compute/aai) integration.
```

File: docs/.gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# =====================================
+# Sphinx documentation
+# =====================================
+_build/
+static/*.key
```
Comment (blankdots) [inline docs/.gitignore]: this line might be a miss.

File: README.md (hunk 2 above)
Comment (blankdots) [reply README.md]: also some installation instructions would be in order.

File: docs/.gitignore (hunk 3 above)
Comment (teemukataja) [reply docs/.gitignore]: Will remove `static/*.key`

File: README.md (hunk 2 above)
Comment (teemukataja) [reply README.md]: Will do

File: docs/configuration.rst [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+.. _env:
+
+Environment Variables
+~~~~~~~~~~~~~~~~~~~~~
+
+The values in the configuration file can be overwritten with environment variables using the exact same name
+in all capital letters. For example:
+
+To overwrite the web application port from 8080 to 3000, one set the following environment variable:
```
Comment (blankdots) [inline docs/configuration.rst]:
```suggestion
To overwrite the web application port from ``8080`` to ``3000``, one set the following environment variable:
```

File: oidc_client/utils/utils.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+import aiohttp
+
+from aiohttp import web
+
+from ..config import CONFIG
+from .logging import LOG
+
+
+def ssl_context():
```
Comment (blankdots) [inline oidc_client/utils/utils.py]: We can have a placeholder for the necessary things here even just returning `None` e.g.:
```
import ssl
sslcontext.load_cert_chain(ssl_certfile, ssl_keyfile)
sslcontext = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
sslcontext.check_hostname = False
```

and documenting how people can set up this.

File: oidc_client/utils/utils.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    """Generate a state for authentication request and return the value for use."""
+    LOG.debug('Generate a new state for authentication request.')
+    return str(uuid4())
+
+
+async def get_from_cookies(request, key):
+    """Get a desired value from cookies."""
+    LOG.debug(f'Retrieve value for {key} from cookies.')
+
+    try:
```
Comment (blankdots) [inline oidc_client/utils/utils.py]: this `try` here is somewhat confusing when is `500` raised? What needs to happen for that?

File: oidc_client/utils/utils.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        async with session.get(CONFIG.aai['url_userinfo'], headers=headers) as response:
+            LOG.debug(f'AAI response status: {response.status}.')
+            # Validate response from AAI
+            if response.status == 200:
+                # Parse response
+                result = await response.json()
+                # Check for Bona Fide values
+                ga4gh = result.get('ga4gh', {})
+                if 'AcceptedTermsAndPolicies' in ga4gh:
+                    for accepted_terms in ga4gh["AcceptedTermsAndPolicies"]:
```
Comment (blankdots) [inline oidc_client/utils/utils.py]: We are looking for at least one of  `AcceptedTermsAndPolicies` to be equal to `CONFIG.elixir['bona_fide_value']` ?
The same goes for `ResearcherStatus` ?
Can there be multiple and if one of them is not equal we ignore it ?

File: oidc_client/utils/utils.py (hunk 6 above)
Comment (teemukataja) [reply oidc_client/utils/utils.py]: Unexpected issues with http request

File: oidc_client/utils/utils.py (hunk 7 above)
Comment (blankdots) [reply oidc_client/utils/utils.py]: Maybe somewhere a link to actual specification: https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md

Comment (teemukataja) [reply oidc_client/utils/utils.py]: [AcceptedTermsAndPolicies](https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#ga4ghacceptedtermsandpolicies) must contain `https://doi.org/10.1038/s41431-018-0219-y` to signify that the user has agreed to the Bona Fide contract.

[ResearcherStatus](https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#ga4ghresearcherstatus) must contain `https://doi.org/10.1038/s41431-018-0219-y` to signify that the user has been vouched for by another esteemed researcher for upholding good ethics and intents.

For gaining Bona Fide attribute to services, both of these requirements must be met. It doesn't matter what other objects they may contain.

Comment (blankdots) [reply oidc_client/utils/utils.py]: I was thinking of the use case where we need to check multiple of them at the same time, it changes a lot the logic of that function.

Comment (blankdots) [reply oidc_client/utils/utils.py]: There are two options here:
* leave it as is with the mention in the docs that this has been designed for checking only Bona Fide for now and do an issue on it to modify it
* make the change now

i assume the former would be more suitable.

Comment (teemukataja) [reply oidc_client/utils/utils.py]: The values in the structure we need to check for are
```
{
  "ga4gh": {
    "AcceptedTermsAndPolicies": [
        {
          "value": "string"
        }
    ],
    "ResearcherStatus": [
        {
          "value": "string"
        }
    ]
  }
}
```
So it's `obj obj list obj str` `{{[{""}]}}` and for loop seemed like a good way to go through those. Would it be possible to check it via some kind of schema like in https://github.com/CSCfi/beacon-python/blob/master/beacon_api/utils/validate.py#L67 ?

Comment (blankdots) [reply oidc_client/utils/utils.py]: sure, but preferably in a different issue.

Comment (blankdots) [reply oidc_client/utils/utils.py]: it seems: https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#registered-access that checking those values for registered access is enough. We can have an issue to make this implementation more flexible later down the road, for now this is enough.

File: oidc_client/utils/utils.py (hunk 5 above)
Comment (teemukataja) [reply oidc_client/utils/utils.py]: described in configuration file

**Pull request merged**